[week3] Spring Web으로 ToDo REST API 만들기

caoyu-dev opened the pull request:
**이번 주 목표**

- [x] HelloController 유닛 테스트
- [x] HelloController MockMVC 테스트
- [x] Task 모델 테스트
- [x] TaskService 테스트
- [x] TaskController 유닛 테스트
- [ ] TaskController MockMVC 테스트

**요구사항**

- [ ] 모든 인수 테스트를 통과해야 합니다.
- [ ] 테스트 커버리지 100%를 달성해야 합니다.

----

이번에도 잘 부탁드립니다. :-)

caoyu-dev committed:
feature(AppTest): AppTest 작성

caoyu-dev committed:
feature(HelloController): HelloController 및 해당 Test 생성

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/HelloControllerTest.java:
```diff
@@ -0,0 +1,23 @@
+
+    @Test
+    void sayHello() {
+        assertThat(helloController.sayHello()).isEqualTo(Hello_Friend);
```
좋아요! 실제 Spring을 이용한 MockMvc 테스트도 있지만 이런 간단한 테스트도 유용하죠!

이제 `Task` 관련 테스트를 추가하실 때, 아래의 링크를 참조해서 만들어보시면 좋겠어요!

https://johngrib.github.io/wiki/junit5-nested/

caoyu-dev committed:
refactor(App): DisplayName 추가

caoyu-dev committed:
refactor(HelloControllerTest): @Nested, @DisplayName 추가

caoyu-dev commented:
`@DisplayName`이 작동하지 않고 테스트 결과로 class명을 출력하는 문제가 있어 해당 문제부터 해결하고 진행해야 할 것 같습니다. ;0

+기록) 예를 들어, 아래와 같은 상태
com.codesoom.assignment.controllers.HelloControllerTest.sayHello_메서드는

문제 해결법 제시)
1. Preferences -> Build, Execution, Deployment -> Build Tools -> Gradle -> Run tests using -> IntelliJ IDEA 로 변경 (X)
2. (https://johngrib.github.io/wiki/intellij/#같은-패키지-안에-있는-멀쩡한-클래스를-못-찾는-경우) File - Invalidate Caches / Restart (X)
3. 그레이들 버전 업그레이드, 다운그레이드 (X)
4. 인텔리제이 재설치 (X)
5. git pull 해서 다시 끌고오기... (ing)

caoyu-dev commented:
5. (회사컴) git pull 로 다시 run -> 이 또한 class명 그대로 출력 (X)
6. 설정에 문제가 있다고 판단하고 살폈더니 test source 설정이 잘못되어 있었음 (O)

---
다시 과제하러 갑니다

caoyu-dev committed:
refactor(Task): 생성자 추가

caoyu-dev committed:
feature(TaskControllerTest): list 메서드에 대한 테스트

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,38 @@
+    }
+
+    @Test
+    @DisplayName("기본적으로 생성된 Task 가 하나만 존재한다면, list 메서드로 조회시 Task 의 갯수는 1개이다.")
```
아주 자세히 설명되어있네요! 한가지 `기본적으로 생성된 Task` 가 좀 더 구체적인 설명이면 좋겠어요!

이 설명도 꽤나 구조화시킬 수 있는데요! 아래 링크를 참고하여 개선해보세요!

https://johngrib.github.io/wiki/junit5-nested/

caoyu-dev replied:
기본적으로 생성된... 다시 읽어보니 굉장히 어색하네요. 하나만 생성되어 있다는 의미로 사용했어요!

caoyu-dev committed:
feature(TaskControllerTest): detail 메서드에 대한 테스트

caoyu-dev committed:
refactor(TaskControllerTest): List 메서드 테스트 DisplayName 수정

caoyu-dev committed:
feature(TaskControllerTest): create 메서드에 대한 테스트

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,83 @@
+    @Test
+    @DisplayName("만약 Task 가 하나만 존재한다면, create 메서드로 생성시 Task 의 id 값이 1 증가한 상태로 Tasks 에 추가된다.")
+    void create() {
+        final Long new_size = Task_Id + 1;
```
좋습니다! 
살짝 걸리는 부분이 있다면 `Task_Id + 1` 는 `Task_Id` 가 1인 경우만 테스트가 가능하군요!
new_size는 `taskController.list().size()` 로 `create` 을 하기 전에 수행하는 것이 이전 size를 알 수 있는 바람직한 상황이 아닐까 싶어요! 


혹은 상수 2를 사용해도 무방할 것 같아요!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,83 @@
+    @Test
+    @DisplayName("만약 Task 가 하나만 존재한다면, create 메서드로 생성시 Task 의 id 값이 1 증가한 상태로 Tasks 에 추가된다.")
+    void create() {
+        final Long new_size = Task_Id + 1;
```
좋습니다! 
살짝 걸리는 부분이 있다면 `Task_Id + 1` 는 `Task_Id` 가 1인 경우만 테스트가 가능하군요!
new_size는 `taskController.list().size()` 로 `create` 을 하기 전에 수행하는 것이 이전 size를 알 수 있는 바람직한 상황이 아닐까 싶어요! 


혹은 상수 2를 사용해도 무방할 것 같아요!

caoyu-dev replied:
`final int new_size = taskController.list().size() + 1` 로 설정한 후 `create` 메서드 과정을 1회 진행하고 `final int actual_size = task().list().size()` 를 구해서 `new_size` 와 `actual_size` 를 비교하는 순서로 변경했습니다. 이 형태는 어떻게 생각하세요?

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,83 @@
+
+@DisplayName("TaskController 클래스")
+class TaskControllerTest {
+    private static final Long Task_Id = 1L;
```
상수네요! 별거아니지만 상수는 `TASK_ID` 이런식으로 이름을 변경해주세요! 아래는 오라클 네이밍 컨벤션을 링크로 공유드릴게요!

https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

caoyu-dev replied:
감사합니다!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,83 @@
+
+@DisplayName("TaskController 클래스")
+class TaskControllerTest {
+    private static final Long Task_Id = 1L;
```
상수네요! 별거아니지만 상수는 `TASK_ID` 이런식으로 이름을 변경해주세요! 아래는 오라클 네이밍 컨벤션을 링크로 공유드릴게요!

https://www.oracle.com/java/technologies/javase/codeconventions-namingconventions.html

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,83 @@
+    class Detail {
+        @Nested
+        @DisplayName("클라이언트가 요청한 Task 의 id 가 존재하면")
+        class valid_id {
```
DisplayName이 있어서 class의 이름은 별거 아니지만! 한번 https://johngrib.github.io/wiki/junit5-nested/ 다음 링크를 따라서 이름 앞에 접두어도 붙여보고 DisplayName의 의미와 비슷하게 만들어보아요!

caoyu-dev committed:
feature: Service updateTask 로직 변경 및 update 메서드에 대한 테스트

caoyu-dev committed:
feature(TaskControllerTest): delete 메서드에 대한 테스트

caoyu-dev committed:
refactor(TaskControllerTest): create 메서드 변수 수정
final int new_size = taskController.list().size() + 1 로 설정한 후
create 메서드 과정을 1회 진행하고
final int actual_size = task().list().size() 를 구해서
new_size 와 actual_size 를 비교하는 순서로 변경

caoyu-dev committed:
refactor: 오라클 네이밍 컨벤션
Task_Id -> TASK_ID 형식으로 변경

caoyu-dev committed:
feature(TaskTest): Task 에 대한 테스트

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/models/TaskTest.java:
```diff
@@ -0,0 +1,51 @@
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@DisplayName("TaskTest 메서드는")
```
TaskTest 메서드 라는 단어가 이질적입니다. 아마도 이곳에는 
Task 클래스를 나타낼 것이고, 아래테스트에서 `getId` 가 설명되어야 할 것 같아요!
`@Nested` 를 이용하면 도움이 될 것입니다!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Test
+    @DisplayName("만약 Task 가 하나만 존재한다면, create 메서드로 생성시 Task 의 id 값이 1 증가한 상태로 Tasks 에 추가된다.")
```
`만약 Task가 하나만 존재한다면` 은 필요없겠어요! `new_size`는 Task에 상관없이 1이 증가할테니까요!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,129 @@
+        }
+
+        @Nested
+        @DisplayName("클라이언트가 수정한 Task 의 Title 과")
```
`update메소드는 클라이언트가 수정한 Task 의 Title 과 전달한 Task 의 Title 은 같아야 한다.` 
은 문장이 어색합니다!! 어떤 컨텍스트에서 update가 실행되었을 때 무엇을 리턴하는지가 기술되면 좋겠습니다.

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,129 @@
+        }
+
+        @Nested
+        @DisplayName("클라이언트가 수정한 Task 의 id 값이 없으면")
```
클라이언트라는 단어는 필요없어요. 만약에 그렇다면 모든 테스트에 클라이언트라는 단어가 필요하겠지요!!

또한 아직 Task는 수정이 수행되지 않았어요! 수정된 Task의 id를 찾지못했다는 것처럼 느껴져요!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/controllers/TaskControllerTest.java:
```diff
@@ -0,0 +1,129 @@
+    }
+
+    @Test
+    @DisplayName("Delete 메서드는 클라이언트가 요청한 Task 의 id 를 삭제한다.")
```
이것도 계층형으로 바꿔볼 수 있겠어요!

caoyu-dev committed:
refactor(TaskTest): @Nested 를 이용한 계층형으로 수정

caoyu-dev committed:
refactor(TaskControllerTest): @DisplayName 수정 및 @Nested 를 이용한 계층형 구조로 변경

caoyu-dev committed:
feature(TaskServiceTest): TaskService 에 대한 테스트

caoyu-dev committed:
package-lock.json: lockfileVersion = 2

caoyu-dev committed:
refactor(HelloControllerWebTest): class 명 변경 및 DisplayName 설정

caoyu-dev committed:
feature(HelloControllerWebTest): HelloController 에 대한 MockMVC 테스트

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/models/TaskTest.java:
```diff
@@ -4,8 +4,9 @@
 
 import static org.assertj.core.api.Assertions.assertThat;
 
+@DisplayName("TaskTest_클래스의")
```
테스트하는 대상을 적는 것이 좋습니다! 
`TaskTest` 도 말이 되지만 `Task` 를 테스트하기에 `Task` 로 적어주는 것이 좋습니다!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,132 @@
+    @Nested
+    class update_메서드는 {
+        @Nested
+        class 만약_해당_id가_있으면 {
```
만약은 제거해주셔도 될 거에요! 항상 Context는 만약이니까요!

해당 id란 무엇을 말하는 걸까요?! 

아마도 변경하고자하는 id나 udpqte를 요청한 id겠죠! 그것을 고려해서 적어주시면 좋겠어요!

ssisksl77 commented on app/src/test/java/com/codesoom/assignment/application/TaskServiceTest.java:
```diff
@@ -0,0 +1,132 @@
+            }
+        }
+    }
+}
```
EOF에는 newline을 추가해주세요!

caoyu-dev committed:
feature: TaskControllerWeb 에 대한 테스트 작성 및 기타 내용 수정
TaskService 에서 updateTask 내용 수정
Task 에서 생성자 추가

caoyu-dev committed:
refactor(TaskServiceTest): EOP 에 newline 추가

caoyu-dev committed:
refactor(TaskServiceTest): class name 수정

caoyu-dev committed:
refactor(TaskTest): DisplayName 수정

caoyu-dev commented:
`update`, `create`, `delete` 메서드에 대한 MockMVC 테스트를 진행하지 못하고 마쳤다.

해당 부분은 아샬님 3회차 피드백 강의로 보충하도록 한다. 

---------------

다음 주도 잘부탁드립니다!

ssisksl77 commented:
한주간 고생많으셨습니다!
